keps/173-headless-service-per-replica/README.md

```diff
@@ -0,0 +1,328 @@
+
+### Goals
+
+- Each group of leader-worker replicas can have a dedicated service
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Providing an API to configure service per replicas.
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a user running a large scale LWS job, I want to have each leader-worker group with its own service.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+```golang
+type LeaderWorkerSetSpec struct {
+ // SubdomainPolicy determines the policy that will be used when creating
+ // the headless service
+ SubdomainPolicy SubdomainPolicy `json:"subdomainPolicy,omitempty"`
+}
+
+type NetworkConfig struct {
+  SubdomainPolicy SubdomainPolicy
+}
+type SubdomainPolicy string
+const (
+  // SubdomainShared will create a single headless service that all replicas 
+  // will share
+  SubdomainShared SubdomainPolicy = "Shared"
+  // SubdomainUniquePerReplica will create a headless service for each
+  // leader-worker group
+  SubdomainUniquePerReplica SubdomainPolicy = "UniquePerReplica"
+)
+```
+
+### Implementation
+
+With HeadlessServicePerReplica true, we will create a headless service per replica.
```

Comment:
maybe we create a headless service per statefulset

Reply:
So the leader will have one, while the workers will have a different one?

Reply:
yes, ideally we only need the headless service to generate the dns record

Reply:
What are the benefits of creating a new one per statefulset instead? Creating a new one per sts creates more headless services

Reply:
The headless service name will be part of the fully qualified host name of the pod, so how many and what name we give the service matters. It is also why I am suggesting that the API to be a knob that decides on the subdomain policy because this is the user facing side effect. 

The pod hostname is `<pod-name>.<svc-name>`

So if the new "policy" is `UniquePerReplica`, then each replica including the leader should be part of the same headless service so that the middle part above is common among the pods of a replica. Example:

Replica 0:
- my-lws-0.my-lws-0
- my-lws-0-1.my-lws-0

Replica 1:
- my-lws-1.my-lws-1
- my-lws-1-1.my-lws-1


If we create a headless service per sts including a special one for leaders, then we can't achieve the above since the leader will have a different subdomain name. Here is what it will look like:

Replica 0:
- my-lws-0.my-lws
- my-lws-0-1.my-lws-0

Replica 1:
- my-lws-1.my-lws
- my-lws-1-1.my-lws-1

I am fine with the above, but what would we call such a domain name "policy"? it wouldn't be `UniquePerReplica`.

Reply:
I prefer the `uniquePerReplica` option. It creates less headless services, and it makes sense for the leader and worker to have the subdomain in common. 

Plus, implementing a headless service per sts will require adding a label or annotation in order for the headless service to only select the workers (we can use `worker-index=0` to only select the leaders)

Reply:
Thinking about it a bit more, creating one exclusively for the leaders would allow for easier transition, since it doesn't require a change to the LWS_LEADER_ADDRESS environment variable. The leader address would still be `fmt.Sprintf("%s-%s.%s.%s", lwsName, groupIndex, lwsName, pod.ObjectMeta.Namespace)` https://github.com/kubernetes-sigs/lws/blob/main/pkg/utils/pod/pod_utils.go#L119

Reply:
I am just struggling to suggest a name that clearly indicates the semantics, what would we call this policy?

Reply:
> I prefer the uniquePerReplica option. It makes sense for the leader and worker to have the subdomain in common.

 +1, or it's weird I think, unless we have other use cases like visiting the workers only.

Reply:
But again, what would you call that policy? it is certainly not `UniquePerReplica` if we do a headless service per sts.

Reply:
What about `uniquePerStatefulSet`?

Reply:
While statefulset is known as the implementation API, it is not a "concept" in the LWS API. The concepts we have are: replica/group, leader, worker etc.

Reply:
What about `uniquePerLeaderAndWorkerReplicas`? It is quite long but I think gets the idea across.

Reply:
`UniquePerReplicaWorkers` maybe?

Reply:
What about `leaderSharedWorkerDedicated` vs `shared`.

Reply:
Of the three I prefer `leaderSharedWorkerDedicated`

Reply:
`LeadersSharedWorkersDedicated`